cilium/CFP-30235-xds-adapter.md

```diff
@@ -0,0 +1,204 @@
+
+## Proposal
+
+### Logic
+
+The introduction of this adapter would mean that Cilium could accept endpoints
+from more than one source. Cilium could be configured to read exclusively from
+Kubernetes (default), xDS, or both, depending on the use case.
+
+When both data sources were in use, precedence in any conflicts would be given
+to data received directly from Kubernetes. As a concrete example, it's possible
+that the same Service could be represented in both Kubernetes and xDS. If that
+were the case, we'd give precedence to the Service from Kubernetes APIs and drop
+the one from xDS, likely surfacing some kind of warning along the way.
+
+### Architecture
+
+This xDS adapter would be similar conceptually to KVStore. Essentially this
+would be another backend that Cilium could get data from. Similar to how Cilium
+can read from Kubernetes APIs and/or KVStore today, we would add a third option
+for xDS as a data source, and develop a shared interface that worked across all
+data sources.
+
+### API Mapping
+
+To keep the initial work as focused as possible, I propose using existing xDS
+APIs to model ClusterIP routing. In the future, we may want to consider either
+additions to those existing xDS APIs or new xDS APIs to cover the entirety of
+Cilium functionality. To start, it seems highly valuable to focus on the areas
+where existing xDS APIs overlap with existing Cilium capabilities.
+
+This shows how we could use existing xDS APIs to represent Cilium capabilities
+for ClusterIP routing:
+
+| Cilium Field | Origin xDS | Comments |
+| - | - | - |
+| Frontend.Address | `filter_chain_match.prefix_ranges` | |
+| Frontend.Port | `filter_chain_match.destination_port` | |
+| Frontend.Protocol | `cluster.metadata.protocol` | |
+| Backend[*].FEPortName | `cluster.metadata.port_name` | |
+| Backend[*].NodeName | ? | TBD - This is only used for logic to determine if an endpoint is local. Instead of sending this through xDS for every endpoint we could just look at the CIDR configured for each Cilium instance. |
+| Backend[*].Addr, Backend[*].Port | `lb_endpoints.endpoint.address.socket_address` | |
+| Backend[*].State | `lb_endpoints.health_status` | |
+| Backend[*].Preferred | N/A | Not used in Kubernetes |
+| Backend[*].Weight | `lb_endpoints.endpoint.load_balancing_weight` | |
+| Type | N/A | Out of scope - only focusing on ClusterIP Services |
+| ExternalTrafficPolicy, InternalTrafficPolicy | `cluster.load_balancing_policy` | Defined in the typed struct `type.googleapis.com/cilium_io_traffic_policy` |
+| NatPolicy | N/A | Not used in Kubernetes |
+| SessionAffinity | `tcpproxy.hash_policy` | |
+| SessionAffinityTimeout | ? | TBD - This represents the maximum session stickiness time. There is no natural place for this in existing xDS APIs. |
+| HealthCheckNodePort | N/A | Out of scope - only focusing on ClusterIP Services |
+| Name | `cluster.metadata.service_name` | |
+| Namespace | `cluster.metadata.service_namespace` | |
+
```

Comment:
Should we list the https://github.com/istio/ztunnel/blob/master/proto/workload.proto as an alternative per @louiscryan's comment on slack?

Reply:
Good question, added a "key question" about this + a reference to WDS.